src/core/lib/http/httpcli.cc

```diff
@@ -41,17 +41,59 @@
 #include "src/core/lib/iomgr/tcp_client.h"
 #include "src/core/lib/slice/slice_internal.h"
 
-struct internal_request {
+namespace {
+
+struct InternalRequest {
+  InternalRequest(const grpc_slice& request_text,
+                  grpc_httpcli_response* response,
+                  grpc_resource_quota* resource_quota, const char* host,
+                  const char* ssl_host_override, grpc_millis deadline,
+                  const grpc_httpcli_handshaker* handshaker,
+                  grpc_closure* on_done, grpc_httpcli_context* context,
+                  grpc_polling_entity* pollent, const char* name)
+      : request_text(request_text),
+        resource_quota(resource_quota),
+        host(gpr_strdup(host)),
+        ssl_host_override(gpr_strdup(ssl_host_override)),
+        deadline(deadline),
+        handshaker(handshaker),
+        on_done(on_done),
+        context(context),
+        pollent(pollent) {
+    grpc_http_parser_init(&parser, GRPC_HTTP_RESPONSE, response);
+    grpc_slice_buffer_init(&incoming);
+    grpc_slice_buffer_init(&outgoing);
+    grpc_iomgr_register_object(&iomgr_obj, name);
+  }
+
+  ~InternalRequest() {
+    grpc_http_parser_destroy(&parser);
+    if (addresses != nullptr) {
+      grpc_resolved_addresses_destroy(addresses);
+    }
+    if (ep != nullptr) {
+      grpc_endpoint_destroy(ep);
+    }
+    grpc_slice_unref_internal(request_text);
+    gpr_free(host);
+    gpr_free(ssl_host_override);
+    grpc_iomgr_unregister_object(&iomgr_obj);
+    grpc_slice_buffer_destroy_internal(&incoming);
+    grpc_slice_buffer_destroy_internal(&outgoing);
+    GRPC_ERROR_UNREF(overall_error);
+    grpc_resource_quota_unref_internal(resource_quota);
+  }
+
   grpc_slice request_text;
   grpc_http_parser parser;
-  grpc_resolved_addresses* addresses;
-  size_t next_address;
-  grpc_endpoint* ep;
+  grpc_resolved_addresses* addresses = nullptr;
+  size_t next_address = 0;
+  grpc_endpoint* ep = nullptr;
   grpc_resource_quota* resource_quota;
   char* host;
```

Comment:
How about changing these `char*` fields to `std::string`?

Reply:
Done.